From: Richard Acayan <mailingradian@gmail.com>
Date: Tue, 19 Sep 2023 21:36:45 -0400
Subject: [PATCH] drm/panel: samsung-sofef00-bonito: move dsi commands to
 enable function

DSI commands are no longer allowed in the prepare function. Move it to a
new enable function so that the panel can successfully turn on.

This is not a confirmed bug on this panel, but was discovered on the
S6E3FA7 panel.

Signed-off-by: Richard Acayan <mailingradian@gmail.com>
---
 .../gpu/drm/panel/panel-samsung-sofef00-bonito.c   | 39 +++++++++-------------
 1 file changed, 16 insertions(+), 23 deletions(-)

diff --git a/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c b/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
index 333265b..e88227d 100644
--- a/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
+++ b/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
@@ -37,8 +37,22 @@ static void sofef00_reset(struct sofef00 *ctx)
 	usleep_range(10000, 11000);
 }
 
-static int sofef00_on(struct sofef00 *ctx)
+static int sofef00_prepare(struct drm_panel *panel)
+{
+	struct sofef00 *ctx = to_sofef00(panel);
+
+	if (ctx->prepared)
+		return 0;
+
+	sofef00_reset(ctx);
+
+	ctx->prepared = true;
+	return 0;
+}
+
+static int sofef00_enable(struct drm_panel *panel)
 {
+	struct sofef00 *ctx = to_sofef00(panel);
 	struct mipi_dsi_device *dsi = ctx->dsi;
 	struct device *dev = &dsi->dev;
 	int ret;
@@ -75,28 +89,6 @@ static int sofef00_on(struct sofef00 *ctx)
 	return 0;
 }
 
-static int sofef00_prepare(struct drm_panel *panel)
-{
-	struct sofef00 *ctx = to_sofef00(panel);
-	struct device *dev = &ctx->dsi->dev;
-	int ret;
-
-	if (ctx->prepared)
-		return 0;
-
-	sofef00_reset(ctx);
-
-	ret = sofef00_on(ctx);
-	if (ret < 0) {
-		dev_err(dev, "Failed to initialize panel: %d\n", ret);
-		gpiod_set_value_cansleep(ctx->reset_gpio, 1);
-		return ret;
-	}
-
-	ctx->prepared = true;
-	return 0;
-}
-
 static int sofef00_disable(struct drm_panel *panel)
 {
 	struct sofef00 *ctx = to_sofef00(panel);
@@ -168,6 +160,7 @@ static int sofef00_get_modes(struct drm_panel *panel,
 
 static const struct drm_panel_funcs sofef00_panel_funcs = {
 	.prepare = sofef00_prepare,
+	.enable = sofef00_enable,
 	.disable = sofef00_disable,
 	.unprepare = sofef00_unprepare,
 	.get_modes = sofef00_get_modes,
